src/relax/transform/lazy_transform_params.cc

```diff
@@ -0,0 +1,277 @@
+    }
+
+    auto output = Downcast<Function>(ExprMutator::VisitExpr_(node.get()));
+    plan_.reset();
+    return output;
+  }
+
+  void VisitBinding(const Binding& binding) override {
+    ExprMutator::VisitBinding(binding);
+    GenerateSetOutputCalls(binding->var, [this](const auto& fset_output_call) {
+      builder_->Emit(fset_output_call, "_void");
+    });
+  }
+
+ private:
+  template <typename Callback>
+  void GenerateSetOutputCalls(const Var& var, Callback callback) {
+    if (plan_.has_value()) {
+      if (auto it = plan_->output_lookup.find(var); it != plan_->output_lookup.end()) {
+        for (auto output_index : it->second) {
+          callback(
+              Call(plan_->fset_output, {PrimValue(IntImm(DataType::Int(64), output_index)), var}));
+        }
+      }
+    }
+  }
+
+  struct FunctionPlan {
+    std::unordered_map<Var, std::vector<size_t>, ObjectPtrHash, ObjectPtrEqual> output_lookup;
+    Expr fset_output;
+  };
+  std::optional<FunctionPlan> plan_;
+};
+}  // namespace
+
+Function WithLazyInputs(Function func) {
+  LazyInputMutator mutator;
+
+  func = Downcast<Function>(mutator.VisitExpr(func));
+  func = Downcast<Function>(EliminateCommonSubexpr(func));
+  func = Downcast<Function>(RemoveAllUnused(func));
+  return func;
+}
+
+Function WithLazyOutputs(Function func) {
+  LazyOutputMutator mutator;
+
+  func = Downcast<Function>(mutator.VisitExpr(func));
+  return func;
+}
+
+namespace transform {
+
+Pass LazyGetInput() {
+  auto pass_func = [](Function func, IRModule, PassContext) -> Function {
+    if (!func->GetAttr<String>(tvm::attr::kGlobalSymbol).defined()) {
+      return func;
+    }
+    return WithLazyInputs(func);
+  };
+  return CreateFunctionPass(/*pass_function=*/pass_func,
+                            /*opt_level=*/0,
+                            /*pass_name=*/"LazyGetInput",
+                            /*required=*/{});
+}
+
+TVM_REGISTER_GLOBAL("relax.transform.LazyGetInput").set_body_typed(LazyGetInput);
+
+Pass LazySetOutput() {
+  auto pass_func = [](Function func, IRModule, PassContext) -> Function {
+    if (!func->GetAttr<String>(tvm::attr::kGlobalSymbol).defined()) {
+      return func;
+    }
+    return WithLazyOutputs(func);
+  };
+  return CreateFunctionPass(/*pass_function=*/pass_func,
+                            /*opt_level=*/0,
+                            /*pass_name=*/"LazySetOutput",
+                            /*required=*/{});
+}
+
+TVM_REGISTER_GLOBAL("relax.transform.LazySetOutput").set_body_typed(LazySetOutput);
+
+// Pass LazyTransformParams() {
+//   auto pass_func = [](Function func, IRModule, PassContext) -> Function {
+//     LazyInput mutator;
+//     return Downcast<Function>(mutator(func));
+//   };
+//   return CreateFunctionPass(/*pass_function=*/pass_func,
+//                             /*opt_level=*/0,
+//                             /*pass_name=*/"MutateOpsForTraining",
+//                             /*required=*/{});
+// }
```

Comment:
I doubt we want to keep this much commented-out code.

Reply:
Thank you, and removing.  My intent is to replace the existing `LazyTransformParams` to be in terms of the C++ implementation, but didn't get around to it.